### ISSUE TITLE
What's the coordinate axis direction during back projection?

### ISSUE THREAD

**Issue by ResearcherYan** [jzhzhang/3DAwareNav#8]
Hello, thanks for your great work! I try to understand the back projection code but the coordinate direction confuses me.

As I understand, the back projection of the depth image is based on the pinhole camera model, which is shown in Fig. 1.

Here the point $P=(x_t, y_t, z_t)$ in the real world is projected to the focal plane as $P\prime = (u_t, v_t)$. For $P=(x_t, y_t, z_t)$, its coordinate system origin is $O_2$, the optical center. For $P\prime = (u_t, v_t)$, its origin is $O_0$, the lower left corner of the image. The center of the image $O_1 = (x_c, y_c)$.

**Fig. 1**
![2024-10-03 19 50 48](https://github.com/user-attachments/assets/cbd4841b-f35c-4233-b180-9d8d1d597940)

**Code Block 1**
```python
def get_point_cloud_from_z_t(Y_t, camera_matrix, device, scale=1):
  """Projects the depth image Y into a 3D point cloud.
  Inputs:
      Y is bsxHxW
      camera_matrix
  Outputs:
      X is positive going right
      Y is positive into the image
      Z is positive up in the image
      XYZ is bsxHxWx3
  """
  # grid_x stands for the x coordinate of each pixel
  # grid_z stands for the z coordinate of each pixel
  grid_x, grid_z = torch.meshgrid(torch.arange(Y_t.shape[-1]),
                                  torch.arange(Y_t.shape[-2] - 1, -1, -1))
  # grid_x:
  # 0 1 2 ... W-1
  # 0 1 2 ... W-1
  # ...
  # 0 1 2 ... W-1
  # grid_z:
  # H-1 H-1 ... H-1
  # H-2 H-2 ... H-2
  # ...
  # 0 0 ... 0
  grid_x = grid_x.transpose(1, 0).to(device)  # H x W
  grid_z = grid_z.transpose(1, 0).to(device)  # H x W
  grid_x = grid_x.unsqueeze(0).expand(Y_t.size())  # bs x H x W
  grid_z = grid_z.unsqueeze(0).expand(Y_t.size())  # bs x H x W

  X_t = (grid_x[:, ::scale, ::scale] - camera_matrix.xc) * \
      Y_t[:, ::scale, ::scale] / camera_matrix.f
  Z_t = (grid_z[:, ::scale, ::scale] - camera_matrix.zc) * \
      Y_t[:, ::scale, ::scale] / camera_matrix.f

  # bs x H x W x 3
  XYZ = torch.stack(
      (X_t, Y_t[:, ::scale, ::scale], Z_t), dim=len(Y_t.size()))

  return XYZ
```

In Code Block 1 above, considering the part computing `X_t` and `Z_t`, assume:
```math
\text{camera\_matrix.xc}=x_c, \text{camera\_matrix.zc}=z_c, \text{camera\_matrix.f}=f,
```
Let's just ignore the broadcasting and only focus on a single point $P=(x_t, y_t, z_t)$. Then the corresponding code is equivalent to the following mathematical formula:
```math
x_t = (u_t - x_c) \cdot \frac{y_t}{f},
```
```math
z_t = (v_t - z_c) \cdot \frac{y_t}{f}.
```
Given $P\prime$ is in the upper right part of the image, such that $u_t - x_c > 0, v_t - z_c > 0$. Since $y_t$ represents the depth info, we also have $y_t > 0$. So $x_t > 0, z_t > 0$. Then the coordinate system of $P=(x_t, y_t, z_t)$ should be what Fig.1 shows.

Here comes the question. **Why is the coordinate system exactly the opposite of the code comments? Plus it's even a left-handed coordinate system!** Could you please point out what's wrong? Thank u a lot!

**Comment by ResearcherYan** [comment]
I thought the above paradox was caused by the image plane (focal plane). I was inspired by Fig. 2. Although the image is formed in the image plane, the camera will always invert it to make sure the customer won't see an upside-down image. So here the virtual image plane is considered as the actual image plane, where the pixel coordinate system lies in.

**Fig. 2**
![image](https://github.com/user-attachments/assets/19d8efc0-5ddb-4ac4-a764-f2caf01ea9bf)

With the assumption above, the coordinate axis direction in the code comments still doesn't make sense. Looking from the optical center to the virtual image plane, the coordinate axes of $P$ are as follows: X is positive going right, Y is positive into the image, Z is positive up in the image. 

Given $P\prime$ lies in the top right of $O_3$, now $P$ also lies in the same direction of $O_4$. We still have $u_t - x_c > 0, v_t - z_c > 0$, so $x_t > 0, z_t > 0$ still holds. But Fig. 3 shows that $x_t < 0$. So I guess new paradox appears. Still need help!

**Fig. 3**
![2024-10-03 21 27 57](https://github.com/user-attachments/assets/d82c9547-090c-4c32-bbb0-5110a052a9f5)

**Comment by jzhzhang** [comment]
We borrowed the code directly from the [Object Goal Nav](https://github.com/devendrachaplot/Object-Goal-Navigation/blob/5d76902fe9be821926a1de32557ca9a8dc21d0f5/envs/utils/depth_utils.py#L129).  And the habitat coordinate documentation can be found in the [link](https://github.com/facebookresearch/habitat-sim/pull/1498)

I’m not entirely sure I fully understand your question. From the comment, Z is defined as `positive upwards in the image`, and in `block 1`, the code uses a descending order, so grid_z becomes `negative downwards in the image`. Therefore, this follows a typical pinhole camera model, which I believe aligns with your equation.”

Please let me know if I have overlooked any of your points.

**Comment by ResearcherYan** [comment]
I'm sorry that maybe I didn't make myself clear. My straight question is: according to the equation, $x_t$ and $z_t$ should be both greater than 0. So the coordinate system for $P$ should be the blue line in Fig. 1. However, it's a left-handed coordinate system. What's more, it's against the code comment. X is positive going left, Y is positive going out of the image, Z is positive going down. **Why does this paradox happen?**

**Comment by jzhzhang** [comment]
It's indeed a left-hand coordinate.  The code,
```grid_x, grid_z = torch.meshgrid(torch.arange(Y_t.shape[-1]), torch.arange(Y_t.shape[-2] - 1, -1, -1))```, generate `grid_x` and `grid_z`. Here, the `grid_x` increases from 0 to W (use W to represent Y_t.shape[-1]), pointing from left to right. And `grid_z` decreases from H-1 (H indicates Y_t.shape[-1]) to 0 (bottom of the image). This means higher `grid_z` values correspond to the top of the image. Considering that the x-axis and y-axis of Pixel Coordinate System, start from left to right and up to down, respectively.

Specifically, when grid_z > camera_matrix.zc, the term (grid_z - camera_matrix.zc) is positive, indicating positions above the camera's vertical center.
Conversely, when grid_z < camera_matrix.zc, the term is negative, indicating positions below the camera's vertical center.

**Comment by ResearcherYan** [comment]
Thanks for your patience. I think I get it now. If it's a left-handed coordinate system, then everything makes sense. The comment says: "Y is positive into the image". It actually means "Y is positive into the virtual image plane", shown in Fig.4. This makes a left-handed coordinate system.

**Fig. 4**
![2024-10-04 09 36 45](https://github.com/user-attachments/assets/d43b1569-3a0a-4901-93c3-281900ca9e2a)

**Comment by jzhzhang** [comment]
Great! I will close this issue. Feel free to reopen it if you have further questions.
